tests/topotests/pim-basic/test_pim.py

```diff
@@ -0,0 +1,142 @@
+# OF THIS SOFTWARE.
+#
+
+"""
+test_pim.py: Test pim
+"""
+
+import os
+import sys
+import pytest
+
+CWD = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(os.path.join(CWD, '../'))
+
+# pylint: disable=C0413
+from lib import topotest
+from lib.topogen import Topogen, TopoRouter, get_topogen
+from lib.topolog import logger
+
+from mininet.topo import Topo
+
+class PIMTopo(Topo):
+    def build(self, *_args, **_opts):
+        "Build function"
+        tgen = get_topogen(self)
+
+        for routern in range(1, 3):
+            tgen.add_router('r{}'.format(routern))
+
+        # r1 <- sw1 -> r2
+        sw = tgen.add_switch('sw1')
+        sw.add_link(tgen.gears['r1'])
+        sw.add_link(tgen.gears['r2'])
+
+
+def setup_module(mod):
+    "Sets up the pytest environment"
+    tgen = Topogen(PIMTopo, mod.__name__)
+    tgen.start_topology()
+
+    # For all registered routers, load the zebra configuration file
+    for rname, router in tgen.routers().iteritems():
+        router.load_config(
+            TopoRouter.RD_ZEBRA,
+            os.path.join(CWD, '{}/zebra.conf'.format(rname))
+        )
+        router.load_config(
+            TopoRouter.RD_PIM,
+            os.path.join(CWD, '{}/pimd.conf'.format(rname))
+        )
+
+    # After loading the configurations, this function loads configured daemons.
+    tgen.start_router()
+    for rname, router in tgen.routers().iteritems():
+        router.run("vtysh -f {}".format(os.path.join(CWD, '{}/frr.conf'.format(rname))))
+
+    #tgen.mininet_cli()
+
+def teardown_module(mod):
+    "Teardown the pytest environment"
+    tgen = get_topogen()
+
+    # This function tears down the whole topology.
+    tgen.stop_topology()
+
+
+def test_pim_send_mcast_stream():
+    logger.info("Establish a Mcast stream from r2->r1 and then ensure S,G created")
+
+    tgen = get_topogen()
+
+    r2 = tgen.gears['r2']
+    r1 = tgen.gears['r1']
+
+    # Let's establish a S,G stream from r2 -> r1
+    CWD = os.path.dirname(os.path.realpath(__file__))
+    out2 = r2.run("{}/mcast-tx.py --ttl 5 --count 5 --interval 10 229.1.1.1 r2-eth0 > /tmp/bar".format(CWD))
+
+    # Let's see that it shows up and we have established some basic state
+    out1 = r1.vtysh_cmd("show ip pim upstream json", isjson=True)
+
+    sg = out1['229.1.1.1']['10.0.20.2']
+    assert sg['firstHopRouter'] == 1
+    assert sg['joinState']  == "NotJoined"
+    assert sg['regState'] == "RegPrune"
+    assert sg['inboundInterface'] == "r1-eth0"
+    #tgen.mininet_cli()
+
+
+def test_pim_igmp_report():
+    logger.info("Send a igmp report from r2-r1 and ensure *,G created")
+
+    tgen = get_topogen()
+
+    r2 = tgen.gears['r2']
+    r1 = tgen.gears['r1']
+
+    # Let's send a igmp report from r2->r1
+    CWD = os.path.dirname(os.path.realpath(__file__))
+    out2 = r2.run("{}/mcast-rx.py 229.1.1.2 r2-eth0 &".format(CWD))
+
+    out1 = r1.vtysh_cmd("show ip pim upstream json", isjson=True)
+    starg = out1['229.1.1.2']['*']
```

Comment:
It is not that hard... It take only a few lines of code to achieve that.

Here is an example with your code.

Before:
```py
    out1 = r1.vtysh_cmd("show ip pim upstream json", isjson=True)

    sg = out1['229.1.1.1']['10.0.20.2']
    assert sg['firstHopRouter'] == 1
    assert sg['joinState']  == "NotJoined"
    assert sg['regState'] == "RegPrune"
    assert sg['inboundInterface'] == "r1-eth0"
```

After:
```py
    # About line 107
    out1 = r1.vtysh_cmd("show ip pim upstream json", isjson=True)

    # Changed those 4 expected lines with a JSON data structure
    expected = {
            "229.1.1.1": {
                "10.0.20.2": {
                    "firstHopRouter": true,
                    "inboundInterface": "r1-eth0",
                    "joinState": "NotJoined",
                    "regState": "RegPrune",
                    }
                }
            }
    assert topotest.json_cmp(out1, expected) is None, 'failed to converge pim'
```

Now instead of getting an error:
```
>       sg = out1['229.1.1.1']['10.0.20.3']
E       KeyError: '10.0.20.3'
```

or

```
>       assert sg['firstHopRouter'] == 2
E       assert True == 2
```

We get:
```
>       assert topotest.json_cmp(out1, expected) is None, 'failed to converge pim'
E       AssertionError: failed to converge pim
E       assert expected key(s) ['10.0.20.3'] in json["229.1.1.1"] (have [u'10.0.20.2']):
E         --- Expected value
E         +++ Current value
E         @@ -2 +2 @@
E         -    "10.0.20.3": {
E         +    "10.0.20.2": {
```

Or:
```
>       assert topotest.json_cmp(out1, expected) is None, 'failed to converge pim'
E       AssertionError: failed to converge pim
E       assert json["229.1.1.1"]["10.0.20.2"]["joinState"] value is different (
E         --- Expected value
E         +++ Current value
E         @@ -1 +1 @@
E         -"Joined"
E         +"NotJoined")
```